### DISCUSSION TITLE
CI/CD - package version from git tag

### DISCUSSION THREAD

**Discussion by FObersteiner** [imk-toolkit/imk-toolkit#53] [Q&A]
I sort-of got hooked by the CI/CD stuff... While contributing to another project, I was wondering if one can take the latest git tag as reference and set the package version based on that.
- I think this is implemented for imktk, right?
- for poetry, I found that one can use `poetry version $(git describe --tags --abbrev=0)` to set the version from the latest git tag to the pyproject.toml file
- How is this done for imktk - I found [this part](https://github.com/imk-toolkit/imk-toolkit/blob/088e0271c1257bdf06d1539570d8faf937075496/.github/workflows/create-release.yml#L26) in the workflow but how do you do it in poetry?

cheers, Florian

**Comment by ucyo** [comment]
Yes. The moment you start with CI/CD you have to be careful. You might never come back :)
![bye-alice-rabbit-hole](https://user-images.githubusercontent.com/591458/155115871-dbfe0fa6-4656-4dda-9a78-abaa2316f56c.gif)
We also do this, but without the git tag appearing in the version number. The part you linked to is the correct starting point. The whole block is this one

https://github.com/imk-toolkit/imk-toolkit/blob/088e0271c1257bdf06d1539570d8faf937075496/.github/workflows/create-release.yml#L26-L44

Its a four step process:
1. Save poetry version
2. Save master branch tag
3. Compare poetry version and tag. If they don't match kill release
4. Check if it is a prerelease version and save it to variable

Then it moves on and creates the release on Docker + pypi

**Reply by FObersteiner** [reply]
Thanks for the explanation! Seems I missed step 3 / the implications of `run [[ v$POETRY_VERSION == $TAG_VERSION ]]` - guess it takes a bit of usage experience to get accustomed to the syntax used in those config files...

btw. here's [this blogpost](https://www.nicholasnadeau.com/post/2020/8/one-version-to-rule-them-all-keeping-your-python-package-version-number-in-sync-with-git-and-poetry/) where they have the poetry version tagging in the github Action?! What do you think of that approach?

**Reply by ucyo** [reply]
This might be a good option actually... Let me test this on my fork ... maybe it works better and makes it easier to deploy

**Comment by FObersteiner** [comment]
I've also come across this: https://pypi.org/project/setuptools-scm/

Anyone used that before? I have my doubts that it would work nicely in combination with poetry...

**Comment by ucyo** [comment]
Do you wanna look at it and make a PR? 👀

**Reply by FObersteiner** [reply]
@ucyo Actually, I had a brief look at it but results weren't satisfactory to me, based on what I could read up and hack together in like 20 min or so.

Now I had another look at the docs and also found [this post](https://stackoverflow.com/questions/56883909/setuptools-scm-current-version-instead-of-next-version) on stackoverflow. After all, I'm not sure what `setuptools-scm` is good for. It creates the *next version*, e.g. like 
```
>>> python -m setuptools_scm
0.1.8rc3.dev4+gb3dd226.d20220314
```
...but you still need to specify a `version` in the pyproject.toml anyway! So just set the tag with `git tag`, put it into your pyproject.toml with `poetry version $(git describe --tags --abbrev=0)` and you're all set to release. What am I missing?

**Reply by ucyo** [reply]
I did not test this, but wouldn't cause `poetry version $(git describe --tags --abbrev=0)` a versioning problem?
Shouldn't the poetry version be set first and then the tag? And when does the push to the repo happen? I am confused 😓

**Reply by FObersteiner** [reply]
> shouldn't the poetry version be set first and then the tag?

@ucyo ok I see where you're getting at - you're absolutely right :) If you 
- tag, 
- then update the version in pyproject.toml
- then commit & push commit + tag to remote

...*then* the tagged version would not contain the commit that has the updated pyproject.toml as it was done after the tagging, i.e. the version there would be wrong. In the blog post I linked, the poetry version is updated in CD (triggered if you push a new tag). So after that, you'd have to pull from your remote so that you have the updated pyproject.toml locally as well. If I remember correctly, we shortly discussed this during our last Python group meeting.

Back to `setuptools_scm`, I still don't see where that fits in :dango:

**Reply by ucyo** [reply]
Exactly... What I mean. But I might have found a solution

**Comment by ucyo** [comment]
How about this solution? We simply use labels on PR if this will trigger a release 🗡️ 
Each PR should have a label with `major`, `minor` or `patch`. The version bumps are automatically generated on merge by CI/CD. 
Additionally we have a second label for `stable` or `release-candidate`. If these labels are present, the release will be triggered.

There is a blog post using this method here: https://blog.pusher.com/continuous-releases-github-actions/

Any thoughts?
